### ISSUE TITLE
cell-var-from-loop spurious warning

### ISSUE THREAD

**Issue by cdyson37** [PyCQA/pylint#1934]
I think this is a spurious warning. In the below, `s` is not defined in any loop.

### Steps to reproduce
    $ cat > bug.py
    # pylint: disable=invalid-name,missing-docstring
    
    for z in [0]:
        s = set()
        def f():
            for x in [0]:
                s.add(x)
        f()
    <EOF>

    
    $ pylint bug.py 
    No config file found, using default configuration
    ************* Module bug
    W:  7,12: Cell variable s defined in loop (cell-var-from-loop)

### Current behavior
Triggers cell-var-from-loop

### Expected behavior
No warning

### pylint --version output
    $ pylint --version
    No config file found, using default configuration
    pylint 1.8.2, 
    astroid 1.6.1
    Python 2.7.14 (default, Sep 16 2017, 17:49:51) 
    [GCC 7.3.0]

**Comment by PCManticore** [comment]
While I admit that this looks like a bug in pylint, I wonder though why the `s` is not initialised in the body of the function, which can then return it?

**Comment by cdyson37** [comment]
This is very much reduced down for a real world (non-insane, honest) example :)

**Comment by PCManticore** [comment]
Gotcha, I was just curious. :)

**Comment by cdyson37** [comment]
The original code iterates up and down a tree visiting things and accumulating what it sees. I could tell you more but I'd have to kill you ;)

**Comment by SamyCookie** [comment]
Does https://github.com/PyCQA/pylint/pull/2276 fixed this issue ?

**Comment by PCManticore** [comment]
@SamyCookie Doesn't seem to be.

**Comment by SamyCookie** [comment]
another (fictive) example that just regressed since 2.10 version, I got also a spurious `cell-var-from-loop` warning:

```python
for n in range(10):
    plus_n = lambda: i + n
    foo = plus_n(42) + plus_n(17)
    print(foo)
```
